crates/dbsp/src/operator/trace.rs

```diff
@@ -391,18 +438,51 @@ where
         &self,
         bounds_stream: &Stream<C, TS>,
         retain_key_func: RK,
+    ) -> Stream<C, Spine<<B as SpillableBatch>::Spilled>>
+    where
+        B: Batch<Time = ()> + SpillableBatch + Send,
+        TS: DBData,
+        RK: Fn(&B::Key, &TS) -> bool + Clone + 'static,
+    {
+        self.shard()
+            .spill()
+            .integrate_trace_retain_keys_in_memory(bounds_stream, retain_key_func)
+    }
+
+    /// Similar to
+    /// [`integrate_trace_retain_keys`](`Self::integrate_trace_retain_keys`),
+    /// but applies a retainment policy to values in the trace.
+    #[track_caller]
+    pub fn integrate_trace_retain_values<TS, RV>(
+        &self,
+        bounds_stream: &Stream<C, TS>,
+        retain_value: RV,
+    ) -> Stream<C, Spine<<B as SpillableBatch>::Spilled>>
+    where
+        B: Batch<Time = ()> + SpillableBatch + Send,
+        TS: DBData,
+        RV: Fn(&B::Val, &TS) -> bool + Clone + 'static,
+    {
+        self.shard()
+            .spill()
+            .integrate_trace_retain_values_in_memory(bounds_stream, retain_value)
+    }
+
+    /// Like [`integrate_trace_retain_keys`](Self::integrate_trace_retain_keys),
+    /// but keeps the trace in memory rather than spilling it to storage.  This
+    /// is appropriate for traces that are known to remain small.
```

Comment:
the problem with traces is that they survive from one clock cycle to the next, unlike other streams which are ephemeral.